### PR TITLE
Unit tests for Http3 xx status code based injection vulnerability

### PR DESCRIPTION
**Pull request by ehizman**

### CODE REVIEW

File: src/main/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjection.java [hunk 1]
```diff
@@ -81,10 +81,11 @@ public ResponseEntity<?> getVulnerablePayloadLevel1(
     // Payloads:
     // 1. Protocol other than http can be used e.g. ftp://ftp.dlptest.com/ also
     // 2. "//facebook.com"
+
     @AttackVector(
             vulnerabilityExposed = {VulnerabilityType.OPEN_REDIRECT_3XX_STATUS_CODE},
             description =
-                    "OPEN_REDIRECT_QUERY_PARAM_DIRECTLY_ADD_TO_LOCATION_HEADER_IF_NOT_HTTP_HTTPS_OR_DOMAIN_IS_SAME")
+                    "OPEN_REDIRECT_QUERY_PARAM_DIRECTLY_ADD_TO_LOCATION_HEADER_IF_NOT_HTTP_HTTPS_WWW_OR_DOMAIN_IS_SAME")
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjection.java:88]: Ehis, we need to update in:

1. https://github.com/SasanLabs/VulnerableApp/blob/master/src/main/resources/i18n/messages.properties
2. https://github.com/SasanLabs/VulnerableApp/blob/master/src/main/resources/i18n/messages_en_US.properties
as well. Actually these strings are locale specific strings so if we don't update them then we will not be able to run the project. I would suggest to please build and run the project after the change to see if everything is running as expected.

**Review by preetkaran20 — CHANGES_REQUESTED** [review]
Please add the strings to message.properties files as well.

File: src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java [hunk 2]
```diff
@@ -0,0 +1,749 @@
...
+            ResponseEntity<?> responseEntity =
+                    http3xxStatusCodeBasedInjection.getVulnerablePayloadLevel6(
+                            requestEntity, redirectUrl);
+            assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.FOUND);
+            assertThat(responseEntity.getHeaders().get(LOCATION_HEADER_KEY))
+                    .contains("https://somedomain.comsomedomain.com");
+            verify(http3xxStatusCodeBasedInjection, times(1))
+                    .getVulnerablePayloadLevel6(requestEntity, "somedomain.com");
+        } catch (URISyntaxException | MalformedURLException exception) {
+            exception.printStackTrace();
```
Comment (preetkaran20) [inline src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java:675]: I think it is better to throw error.

File: src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java [hunk 3]
```diff
@@ -0,0 +1,749 @@
...
+                                    "https://somedomain.com?returnTo=//"
+                                            + (char) 0
+                                            + "www.somedomain.com"));
+            ResponseEntity<?> responseEntity =
+                    http3xxStatusCodeBasedInjection.getVulnerablePayloadLevel5(
+                            requestEntity, redirectUrl);
+            assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
+            assertThat(responseEntity.getHeaders().get(LOCATION_HEADER_KEY)).isEqualTo(null);
+            verify(http3xxStatusCodeBasedInjection, times(1))
+                    .getVulnerablePayloadLevel5(requestEntity, (char) 0 + "//www.somedomain.com");
```
Comment (preetkaran20) [inline src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java:579]: Try using null character present in Constants file.

File: src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java [hunk 4]
```diff
@@ -0,0 +1,749 @@
...
+        verify(http3xxStatusCodeBasedInjection, times(1))
+                .getVulnerablePayloadLevel1("https://www.malicious.com");
+    }
+
+    @Test
+    @DisplayName(
+            "Level 2- test that the returnTo query parameter's value is not added to the Location header when it starts with http")
+    void
+            test_That_ReturnToQueryParameterValue_IsNotAddedToLocationHeader_WhenItStartsWith_Http_Level2() {
+        try {
```
Comment (preetkaran20) [inline src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java:49]: No need to add try/catch

File: src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java [hunk 5]
```diff
@@ -0,0 +1,749 @@
...
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.springframework.http.*;
+
+class Http3xxStatusCodeBasedInjectionTest {
```
Comment (preetkaran20) [inline src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java:17]: Great test coverage @ehizman.

File: src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java (hunk 4 above)
Comment (ehizman) [reply src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java:49]: Okay. Noted

File: src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java (hunk 3 above)
Comment (ehizman) [reply src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java:579]: Okay. Noted

File: src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java [hunk 6]
```diff
@@ -0,0 +1,658 @@
...
+        String redirectUrl = "/";
+        RequestEntity<String> requestEntity =
+                new RequestEntity<>(
+                        HttpMethod.GET, new URI("https://somedomain.com?returnTo=/somedomain.com"));
+        ResponseEntity<?> responseEntity =
+                http3xxStatusCodeBasedInjection.getVulnerablePayloadLevel8(
+                        requestEntity, redirectUrl);
+        assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.FOUND);
+        assertThat(responseEntity.getHeaders().get(LOCATION_HEADER_KEY)).contains("/");
+        verify(http3xxStatusCodeBasedInjection, times(1))
```
Comment (preetkaran20) [inline src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java:655]: as we have called this method at line 651 so this should be giving 1 time invocation. we can ignore this check.

File: src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java [hunk 7]
```diff
@@ -0,0 +1,658 @@
...
+        String redirectUrl = "/somedomain.com";
+        RequestEntity<String> requestEntity =
+                new RequestEntity<>(
+                        HttpMethod.GET, new URI("https://somedomain.com?returnTo=/somedomain.com"));
+        ResponseEntity<?> responseEntity =
+                http3xxStatusCodeBasedInjection.getVulnerablePayloadLevel8(
+                        requestEntity, redirectUrl);
+        assertThat(responseEntity.getStatusCode()).isEqualTo(HttpStatus.OK);
+        assertThat(responseEntity.getHeaders().get(LOCATION_HEADER_KEY)).isEqualTo(null);
+        verify(http3xxStatusCodeBasedInjection, times(1))
```
Comment (preetkaran20) [inline src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java:636]: same we have called this method

**Review by preetkaran20 — APPROVED** [review]
overall looks good. only one cosmetic change.

File: src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java (hunk 7 above)
Comment (ehizman) [reply src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java:636]: Hi @preetkaran20 ,
I need some clarity, should I remove the verify checks?

Comment (preetkaran20) [reply src/test/java/org/sasanlabs/service/vulnerability/openRedirect/Http3xxStatusCodeBasedInjectionTest.java:636]: yes as you have called the method from test case only so checking the verify doesn't give any benefit.

**Comment by preetkaran20** [comment]
Thank you !!!

**Comment by ehizman** [comment]
Thank you Karan!!!

**Pull request merged**